lib/ralph/irc.ex

```diff
@@ -0,0 +1,241 @@
+defmodule Ralph.IRC do
+  ## Configuration
+
+  defmodule RootConfig do
+    defstruct mod: nil, networks: [], hooks: []
+  end
+
+  defmodule NetworkConfig do
+    defstruct name: nil, servers: [], channels: [], nick: nil, hooks: []
+  end
+
+  defmodule ChannelConfig do
+    defstruct name: nil, hooks: []
+  end
+
+  defmacro __using__(_opts) do
+    quote do
+      import Ralph.IRC
+      require Logger
+
+      Module.register_attribute(__MODULE__, :context, persist: true)
+      Module.put_attribute(__MODULE__, :context, %RootConfig{mod: __MODULE__})
+
+      @before_compile unquote(__MODULE__)
+    end
+  end
+
+  defmacro __before_compile__(_env) do
+    quote do
+      def child_spec(opts) do
+        %{id: __MODULE__, start: {__MODULE__, :start_link, [opts]}}
+      end
+
+      def start_link(_opts) do
+        Ralph.IRC.Supervisor.start_link(@context)
+      end
+
+      # on_command?
+      def on_line({network, _}, {prefix, command, params} = message) do
+        Logger.debug("Received message #{inspect message}")
+        Enum.each(@context.hooks, fn hook ->
+          hook_ctx = %{network: network, message: message}
+          apply(__MODULE__, hook, [hook_ctx])
+        end)
+      end
+    end
+  end
+
+  ## Message Hooks
+
+  defmacro on_command(command, handler, param_names) do
+    handler_name = :"_on_command_hook__line_#{__CALLER__.line}"
+
+    quote do
+      def unquote(handler_name)(context, {prefix, cmd, params}) do
+        if cmd == unquote(command) do
+          params =
+            if is_function(unquote(param_names)) do
+              unquote(param_names).(params)
+            else
+              Enum.zip(unquote(param_names), params) |> Enum.into(%{})
+            end
+
+          context = Map.merge(context, params)
+
+          unquote(handler).(context)
+        end
+      end
+
+      add_hook(unquote(handler_name))
+    end
+  end
+
+  defmacro on_join(handler) do
```

Comment:
Figure out a way to remove this quote/unquote dance. Ideally we should be able to define new event hooks doing something like:

```elixir
define_hook, :on_join, "JOIN", handler, [:channel]
```